legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/LegendLanguageService.java

```diff
@@ -229,4 +246,57 @@ public CompletableFuture<List<LegendEntity>> entities(LegendEntitiesRequest requ
             return entities;
         });
     }
+
+    @Override
+    public CompletableFuture<ApplyWorkspaceEditResponse> jsonEntitiesToPureTextWorkspaceEdits(LegendJsonToPureRequest request)
+    {
+        LegendSDLCFeature handler = this.server.getGlobalState().findFeatureThatImplements(LegendSDLCFeature.class).findAny().orElseThrow(() -> new RuntimeException("Could not find feature to convert json to pure files"));
+
+        return this.server.supplyPossiblyAsync(() ->
+        {
+            List<Either<TextDocumentEdit, ResourceOperation>> edits = new ArrayList<>();
+
+            for (String jsonFileUri : request.getJsonFileUris())
+            {
+                int indexOfJsonEntityDirectory = jsonFileUri.indexOf(JSON_ENTITY_DIRECTORY);
+                if (indexOfJsonEntityDirectory != -1)
+                {
+                    try
+                    {
+                        Path path = Paths.get(URI.create(jsonFileUri));
+                        List<String> jsonText = Files.readAllLines(path);
+                        String pureText = "// Converted by Legend LSP from JSON file: " + jsonFileUri.substring(indexOfJsonEntityDirectory + 1) + "\n"
+                                + handler.entityJsonToPureText(String.join("", jsonText));
+
+                        String pureFileUri = jsonFileUri
+                                .replace(JSON_ENTITY_DIRECTORY, "/src/main/pure/")
+                                .replace(".json", ".pure");
+
+                        // rename json file to pure file
+                        edits.add(Either.forRight(new RenameFile(jsonFileUri, pureFileUri)));
+
+                        // add pure content to it
+                        VersionedTextDocumentIdentifier textDocument = new VersionedTextDocumentIdentifier(pureFileUri, null);
+                        TextEdit addPureContent = new TextEdit(
+                                new Range(
+                                        new Position(0, 0),
+                                        new Position(jsonText.size(), 0)
+                                ), pureText);
+                        edits.add(Either.forLeft(new TextDocumentEdit(textDocument, List.of(addPureContent))));
+
+                        this.server.logInfoToClient("Converting JSON protocol to Pure text for: "
+                                + jsonFileUri + " -> " + pureFileUri);
+                    }
+                    catch (Exception e)
+                    {
+                        this.server.logErrorToClient("Failed to convert JSON to pure for: " + jsonFileUri + " - " + e.getMessage());
```

Comment:
I would still recommend logging the exception on the server side so that there's a record of the stack trace.
```suggestion
                        LOGGER.error("Error converting JSON to Pure for {}", jsonFileUri, e);
                        this.server.logErrorToClient("Failed to convert JSON to pure for: " + jsonFileUri + " - " + e.getMessage());
```

Reply:
Every line logged to client is also logged to server - https://github.com/finos/legend-engine-ide-lsp/blob/main/legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/server/LegendLanguageServer.java#L672

Reply:
I would still recommend logging in a way that prints the stack trace in the server logs. The way it is now has the message, but discards a lot of information that would be useful in debugging issues.